src/app/blog/0rtt-api/page.mdx

```diff
@@ -0,0 +1,394 @@
+</Note>
+
+## Normal TLS handshake
+
+<div className="not-prose">
+  <img src="/blog/0rtt-api/1rtt.png" width={1200} height={754} alt="minimal 0rtt interaction" />
+</div>
+
+### ClientHello
+
+The connection is initiated from the client by sending a `ClientHello`. This message contains the TLS protocol version, random data, and the set of supported cipher suites. In addition iroh always includes a set of ALPN strings that identify the application level protocols that the client wishes to speak.
+
+When receiving the `ClientHello`, the server has information to narrow down the set of cryptographic primitives to use. The server endpoint also has a set of ALPNs it supports, so it can downselect those as well.
+
+<Note>
+If the intersection of the ALPNs requested by the client and the ALPNs supported by the server is empty, the handshake will nevertheless continue.
+
+If the intersection of *cipher suites* between client and server is empty however, the handshake will *definitely* fail. The server will not send a `ServerHello`, but just a TLS alert indicating a handshake failure.
+</Note>
+
+### ServerHello
+
+The `ServerHello` message contains a random value as well as a *single* cipher suite that is to be used for further communication. It also contains a *single* ALPN selected by the server to select the application level protocol for the rest of the session - or none if there was no overlap.
+
+So in short, the purpose of the `ServerHello` is to downselect the cipher suite and ALPN, and to contribute randomness for the generation of a unique session specific symmetric key.
+
+The server will send the `ServerHello`, possibly additional encrypted messages for connection setup, and then a `Finished` message.
+
+### Session setup
+
+Once the client has received the `ServerHello`, it has collected all the required information to set up the session for transporting user data. At this point the cipher suite and ALPN are fixed for the rest of the session.
+
+<Note>
+In the case where there is no overlap between client and server ALPNs, in principle it is possible to fall back to a default ALPN, but this is not relevant for iroh connections.
+</Note>
+
+Since the cipher suite is now fixed, the client can now use the client and server random to derive two symmetric keys, one for additional handshake messages and one for application data.
+
+Every message that the client receives after the `ServerHello` is encrypted with the symmetric keys that are now shared knowledge on both sides.
+
+The client processes possible additional packages from the server until it receives the `Finished` message. At that point it validates that it has received all setup messages form the server using a secure checksum contained in the `Finished` message.
```

Comment:
```suggestion
The client processes possible additional packages from the server until it receives the `Finished` message. At that point it validates that it has received all setup messages form the server using a secure checksum contained in the server's `Finished` message.
```
This tripped me over and I had to back-track to understand.  I think with this suggestion it's clearer.